Helm Chart for Kubernetes Deployment

dhanainme opened the pull request:
First iteration for Helm based Torch Serve Installation in EKS.

[x] Helm chart based TS Serve Installation 
[x] EFS Provisioning scripts for PVC
[x] Working End - End example

Next steps in subsequent PRs after 0.2 release:
* Logs / Metrics aggregation
* Autoscaling

This uses a PVC with TS snapshot feature to ensure state of persisted across the pods. The [README](https://github.com/pytorch/serve/tree/k8s/kubernetes) provides a full overview of the Helm Chart.

**Testing**  :  Tested on my EKS cluster. More info in the README

chauhang commented:
@dhanainme The EFS mount is not working on the EKS side. Please add proper troubleshooting steps.

Error Logs/Details:
----------
```
kubectl apply -f templates/efs_pv_claim.yaml
persistentvolumeclaim/model-store-claim created
pod/model-store-pod created
```

But after that status is in pending mode:
```
kubectl get pod
NAME                     READY  STATUS   RESTARTS  AGE
efs-provisioner-1595709188-5f8d89f99f-r6jv2  0/1   Init:0/1  0     2m40s
model-store-pod                0/1   Pending  0     2m16s
```

Getting details of the pod shows following error:

```
Normal  Scheduled  3m57s default-scheduler                   Successfully assigned default/efs-provisioner-1595709188-5f8d89f99f-r6jv2 to ip-192-168-54-247.us-west-2.compute.internal
 Warning FailedMount 114s  kubelet, ip-192-168-54-247.us-west-2.compute.internal Unable to mount volumes for pod "efs-provisioner-1595709188-5f8d89f99f-r6jv2_default(0f051423-ceb6-11ea-addd-0623b987d74e)": timeout expired waiting for volumes to attach or mount for pod "default"/"efs-provisioner-1595709188-5f8d89f99f-r6jv2". list of unmounted volumes=[pv-volume]. list of unattached volumes=[pv-volume efs-provisioner-1595709188-token-6dcfj]

```

Output of pod logs: 
```

kubectl logs pods/efs-provisioner-1595709188-5f8d89f99f-r6jv2

16m         Normal    SuccessfulCreate    replicaset/efs-provisioner-1595709188-5f8d89f99f    Created pod: efs-provisioner-1595709188-5f8d89f99f-r6jv2
16m         Normal    ScalingReplicaSet   deployment/efs-provisioner-1595709188               Scaled up replica set efs-provisioner-1595709188-5f8d89f99f to 1
11m         Warning   FailedMount         pod/efs-provisioner-1595709188-5f8d89f99f-r6jv2     MountVolume.SetUp failed for volume "pv-volume" : mount failed: exit status 32
Mounting command: systemd-run
Mounting arguments: --description=Kubernetes transient mount for /var/lib/kubelet/pods/0f051423-ceb6-11ea-addd-0623b987d74e/volumes/kubernetes.io~nfs/pv-volume --scope -- mount -t nfs fs-f9e33bfc.efs.us-west-2.amazonaws.com:/ /var/lib/kubelet/pods/0f051423-ceb6-11ea-addd-0623b987d74e/volumes/kubernetes.io~nfs/pv-volume
Output: Running scope as unit run-32727.scope.
mount.nfs: Connection timed out
7m14s       Warning   FailedMount         pod/efs-provisioner-1595709188-5f8d89f99f-r6jv2     MountVolume.SetUp failed for volume "pv-volume" : mount failed: exit status 32
Mounting command: systemd-run
Mounting arguments: --description=Kubernetes transient mount for /var/lib/kubelet/pods/0f051423-ceb6-11ea-addd-0623b987d74e/volumes/kubernetes.io~nfs/pv-volume --scope -- mount -t nfs fs-f9e33bfc.efs.us-west-2.amazonaws.com:/ /var/lib/kubelet/pods/0f051423-ceb6-11ea-addd-0623b987d74e/volumes/kubernetes.io~nfs/pv-volume
Output: Running scope as unit run-4472.scope.
mount.nfs: Connection timed out
2m50s       Warning   FailedMount         pod/efs-provisioner-1595709188-5f8d89f99f-r6jv2     MountVolume.SetUp failed for volume "pv-volume" : mount failed: exit status 32
Mounting command: systemd-run
Mounting arguments: --description=Kubernetes transient mount for /var/lib/kubelet/pods/0f051423-ceb6-11ea-addd-0623b987d74e/volumes/kubernetes.io~nfs/pv-volume --scope -- mount -t nfs fs-f9e33bfc.efs.us-west-2.amazonaws.com:/ /var/lib/kubelet/pods/0f051423-ceb6-11ea-addd-0623b987d74e/volumes/kubernetes.io~nfs/pv-volume
Output: Running scope as unit run-8685.scope.
mount.nfs: Connection timed out
43s         Warning   FailedScheduling       pod/model-store-pod                                 pod has unbound immediate PersistentVolumeClaims
38s         Normal    ExternalProvisioning   persistentvolumeclaim/model-store-claim             waiting for a volume to be created, either by external provisioner "example.com/aws-efs" or manually created by system administrator
26s         Warning   FailedMount            pod/efs-provisioner-1595709188-5f8d89f99f-r6jv2     Unable to mount volumes for pod "efs-provisioner-1595709188-5f8d89f99f-r6jv2_default(0f051423-ceb6-11ea-addd-0623b987d74e)": timeout expired waiting for volumes to attach or mount for pod "default"/"efs-provisioner-1595709188-5f8d89f99f-r6jv2". list of unmounted volumes=[pv-volume]. list of unattached volumes=[pv-volume efs-provisioner-1595709188-token-6dcfj]`
```

chauhang commented on kubernetes/README.md:
```diff
@@ -0,0 +1,255 @@
+* eksctl - [Installation](https://docs.aws.amazon.com/eks/latest/userguide/getting-started-eksctl.html)
+* kubectl - [Installation](https://kubernetes.io/docs/tasks/tools/install-kubectl/)
+* helm - [Installation](https://helm.sh/docs/intro/install/)
+
```
@dhanainme Please add install steps for jq

chauhang commented on kubernetes/setup_efs.sh:
```diff
@@ -0,0 +1,39 @@
+MOUNT_TARGET_GROUP_NAME="eks-efs-group"
+
+# Fetch VPC ID / CIDR Block for the EKS Cluster
+CLUSTER_NAME=TorchserveCluster
```
Duplicate of line 4

dhanainme replied:
Fixed.

chauhang commented on kubernetes/templates/eks_cluster.yaml:
```diff
@@ -0,0 +1,15 @@
+  
+nodeGroups:
+  - name: ng-1
+    instanceType: g4dn.xlarge
```
@dhanainme  Please add instructions for what user needs to change to customize these -- choose diff instance type, aws region etc. To make the template general pick values from env variable or add option for generating the templates

dhanainme replied:
Fixed

chauhang commented on kubernetes/values.yaml:
```diff
@@ -0,0 +1,25 @@
+
+efs:
+  region: us-west-2
+  file_system_dns_name: fs-4d0ed148.efs.us-west-2.amazonaws.com
```
file system dns and file system id will be very specific to user's env and we can't have defaults for these

dhanainme replied:
Fixed

dhanainme commented:
@chauhang  

Have added details steps for 
* Installation (jq & others)
* Verification / Expected results for all commands
* Troubleshooting steps for all the steps
* Deletion / Cleanup

Also, 

* Verbose instruction for EFS setup / troubleshooting
* Fixed `setup_efs.sh` to be a bit more verbose / defensive.
* Added,  screenshots for EKS Cluster / EFS / Security group. 

I have tested this in a new EKS cluster / fresh ubuntu 18.04 host.

chauhang commented on kubernetes/values.yaml:
```diff
@@ -0,0 +1,20 @@
+  replicas: 1
+
+persitant_volume:
+  size: 1Gi
```
@dhanainme Where is this size getting used? If user's have lots of models they will need to increase. Add some description in readme for this

dhanainme replied:
Have added a note for templates/efs_pv_claim.yaml & this.

chauhang commented on kubernetes/config.properties:
```diff
@@ -0,0 +1,8 @@
+inference_address=http://0.0.0.0:8080
+management_address=http://0.0.0.0:8081
+NUM_WORKERS=1
+number_of_gpu=1
```
@dhanainme Does the number_of_gpu need to match with the n_gpu value in the TS helm chart?

dhanainme replied:
`n_gpu` would be exposed to TS container by docker. This should be set to `number_of_gpu` in `config.properties` above.

Have added a note

chauhang commented on kubernetes/config.properties:
```diff
@@ -0,0 +1,8 @@
+inference_address=http://0.0.0.0:8080
+management_address=http://0.0.0.0:8081
+NUM_WORKERS=1
```
@dhanainme We will probably need to describe this is a worker insider TorchServe and not the worker nodes on Kubernetes

dhanainme replied:
Done

chauhang commented on kubernetes/values.yaml:
```diff
@@ -0,0 +1,20 @@
+  memory_request: 1Gi
+
+deployment:
+  replicas: 1
```
@dhanainme Mention that the replicas value needs to be <= the total number of worker nodes in the Kubernetes cluster

dhanainme replied:
Done

chauhang commented on kubernetes/README.md:
```diff
@@ -0,0 +1,815 @@
+
+  
+
+  * Delete EKS cluster `eksctl delete cluster -name YOUR_CLUSTER_NAME`
```
@dhanainme It should with `--name` with two hyphens

dhanainme replied:
Done

chauhang commented on kubernetes/README.md:
```diff
@@ -0,0 +1,815 @@
+
+  * Delete EKS cluster `eksctl delete cluster -name YOUR_CLUSTER_NAME`
+  * Delete EFS `aws efs delete-file-system --file-system-id YOUR_EFS_FS_ID`
+  * Delete Security Groups ``aws ec2 delete-security-group --group-id YOUR_SECURITY_GRP_ID` 
```
@dhanainme What is the Security group for? There was no explicit config done during install for the security group. So how will user's get the value of 'YOUR_SECURITY_GRP_ID' for deleting.

dhanainme replied:
This is the output from `setup_efs.sh` https://github.com/pytorch/serve/blob/k8s/kubernetes/setup_efs.sh#L20

dhanainme commented on kubernetes/setup_efs.sh:
```diff
@@ -0,0 +1,56 @@
+# Fetch VPC ID / CIDR Block for the EKS Cluster
+
+echo "Obtaining VPC ID for $CLUSTER_NAME"
+VPC_ID=$(aws eks describe-cluster --name $CLUSTER_NAME --query "cluster.resourcesVpcConfig.vpcId" --output text)
```
This is picked up from AWS_DEFAULT_REGION env. Have confirmed this

dhanainme commented on kubernetes/values.yaml:
```diff
@@ -0,0 +1,20 @@
+# Default values for torchserve helm chart.
+
+torchserve_image: pytorch/torchserve:latest-gpu
```
Done

dhanainme commented on kubernetes/values.yaml:
```diff
@@ -0,0 +1,20 @@
+  management_port: 8081
+  inference_port: 8080
+  pvd_mount: /home/model-server/shared/
+  n_gpu: 1
```
Have added a note.

dhanainme commented on kubernetes/setup_efs.sh:
```diff
@@ -0,0 +1,56 @@
+
+# Fetch VPC ID / CIDR Block for the EKS Cluster
+
+echo "Obtaining VPC ID for $CLUSTER_NAME"
```
Have updated the instruction to `source ./setup_efs.sh` . This would set the env variables so that we can refer at a later time.

dhanainme commented:
@chauhang , Have fixed all the feedback in https://github.com/pytorch/serve/pull/551/commits/eccdc14d2815e05cb5579f93ddf0dc290fa796a3 & https://github.com/pytorch/serve/pull/551/commits/6330eafc3ac5360f32128ce2b61481ee0937193a

dhanainme commented:
@chauhang , Just wanted to followup with you for more feedback / if it can be merged.
